rustPlatform: increase build-speed of `checkPhase` for rust-packages (#82342)
NixOS/nixpkgs: doc/languages-frameworks/rust.section.md

}
```

### Running package tests

When using `buildRustPackage`, the `checkPhase` is enabled by default and runs
`cargo test` on the package to build. To make sure that we don't compile the
sources twice and to actually test the artifacts that will be used at runtime, 
the tests will be ran in the `release` mode by default.

> However, in some cases the test-suite of a package doesn't work properly in the

Reviewer: I think it should be noted explicitly that using `checkType = "debug";` will compile two times -- in release mode for the actual package, and in debug mode for testing (as was happening before this change).

e.g. "NOTE: This will cause two compilations -- one in `release` mode for the package itself, and one in `debug` mode for the tests." or something.

so:

```nix
rustPlatform.buildRustPackage {
  /* ... */
  checkType = "debug";
}
```

#### Tests relying on the structure of the `target/` directory

Some tests may rely on the structure of the `target/` directory. Those tests
are likely to fail because we use `cargo --target` during the build. This means that
the artifacts
[are stored in `target/<architecture>/release/`](https://doc.rust-lang.org/cargo/guide/build-cache.html),
rather than in `target/release/`.

This can only be worked around by patching the affected tests accordingly.

#### Disabling package-tests

In some instances, it may be necessary to disable testing altogether (with `doCheck = false;`):

* If no tests exist -- the `checkPhase` should be explicitly disabled to skip
  unnecessary build steps to speed up the build.
* If tests are highly impure (e.g. due to network usage).

There will obviously be some corner-cases not listed above where it's sensible to disable tests.
The above are just guidelines, and exceptions may be granted on a case-by-case basis.

However, please check if it's possible to disable a problematic subset of the
test suite and leave a comment explaining your reasoning.

> ### Building a package in the `debug` mode

Reviewer: ```suggestion
### Building a package in `debug` mode
```

}
```

### Running package tests

When using `buildRustPackage`, the `checkPhase` is enabled by default and runs
`cargo test` on the package to build. To make sure that we don't compile the

> sources twice and to actually test the artifacts that will be used after that, 

Reviewer 2: ```suggestion
sources twice and to actually test the artifacts that will be used at runtime, 
```

> When using `buildRustPackage`, the `checkPhase` is enabled by default and runs
> `cargo test` on the package to build. To make sure that we don't compile the
> sources twice and to actually test the artifacts that will be used after that, 
> the tests will be ran in the `release`-mode by default.

Reviewer: I think "to actually test the artifacts that will be used" is a bit unnecessary. What else would `cargo test` be testing?

```suggestion
When using `buildRustPackage`, the `checkPhase` is enabled by default and runs
`cargo test` on the package. In order to prevent compiling the source twice, the
tests will be ran in `release` mode by default.
```

Author: > I think "to actually test the artifacts that will be used" is a bit unnecessary. What else would cargo test be testing? 

The build artifacts from the `debug` mode (i.e. without optimizations etc.).

> the tests will be ran in the `release`-mode by default.

Reviewer 2: ```suggestion
the tests will be ran in the `release` mode by default.
```

> However, in some cases the test-suite of a package doesn't work properly in the
> `release` mode. In that case, the mode for `checkPhase` can be changed like this:

Reviewer: "in some cases" followed by "in that case" in the next sentence is too repetitive.

```suggestion
However, in some cases, the test-suite of a package doesn't work properly in the
`release` mode. For these situations, the mode for `checkPhase` can be changed like so:
```

so:

```nix
rustPlatform.buildRustPackage {
  /* ... */
  checkType = "debug";
}
```

> Please note that the code will be compiled twice here: in the `release` mode
> in the `buildPhase` and in the `debug` mode for the `checkPhase`.

Reviewer: One last nit (we dropped "the" from "the `....` mode" everywhere else, should do it here as well):

```suggestion
Please note that the code will be compiled twice here: once in `release` mode
for the `buildPhase`, and again in `debug` mode for the `checkPhase`.
```

> #### Tests relying on the structure of the `target/`-directory
> 
> Some tests may rely on the structure of the `target/`-directory. Those tests
> are likely to fail since we use `cargo --target` during build. This means that
> the artifacts
> [are stored in `target/<architecture>/release/`](https://doc.rust-lang.org/cargo/guide/build-cache.html)
> rather than `target/release/`.

Reviewer: Nits.

```suggestion
#### Tests relying on the structure of the `target/` directory

Some tests may rely on the structure of the `target/` directory. Those tests
are likely to fail because we use `cargo --target` during the build. This means that
the artifacts
[are stored in `target/<architecture>/release/`](https://doc.rust-lang.org/cargo/guide/build-cache.html),
rather than `target/release/`.
```

> rather than `target/release/`.

Reviewer 2: ```suggestion
rather than in `target/release/`.
```

> This can only be circumvented by patching the affected tests accordingly.

Reviewer 2: ```suggestion
This can only be worked around by patching the affected tests accordingly.
```

#### Disabling package-tests

> In some cases it's necessary to disable the tests (which can be done by declaring
> `doCheck = false;`) which is fine in the following cases:

Reviewer: ```suggestion
In some instances, it may be necessary to disable testing with `doCheck = false;`.
Acceptable reasons for doing so include the following:
```

Reviewer 2: @cole-h I would add *altogether*:

> In some instances, it may be necessary to disable testing altogether (with `doCheck = false;`)

* If no tests exist, the `checkPhase` should be explicitly disabled to skip

>   unnecessary build-steps to speed-up the build.

Reviewer 2: ```suggestion
  unnecessary build steps to speed up the build.
```

> * If no tests exist, the `checkPhase` should be explicitly disabled to skip
>   unnecessary build-steps to speed-up the build.
> 
> * If tests are highly impure (e.g. due to heavy network usage), it's also fine
>   disable tests.

Reviewer: ```suggestion
* If no tests exist -- the `checkPhase` should be explicitly disabled to skip
  unnecessary build-steps to speed-up the build.

* If tests are highly impure (e.g. due to network usage).
```

> There are obviously some other corner-cases where it's sensible to disable tests,
> those aren't hard-rules, in the end this is a case-by-case decision.
> 
> Please check however if it's possible to disable a problematic subset of the
> test-suite and leave comment which explains why that's needed.

Reviewer: ```suggestion
There will obviously be some corner-cases not listed above where it's sensible to disable tests.
The above are just guidelines, and exceptions may be granted on a case-by-case basis.

However, please check if it's possible to disable a problematic subset of the
test suite and leave a comment explaining your reasoning.
```

> ### Building a package in the `debug` mode
> 
> By default, `buildRustPackage` will use the `release`-mode for building. If a package
> should be built in the `debug`-mode however, it can be configured like this:

Reviewer: ```suggestion
### Building a package in `debug` mode

By default, `buildRustPackage` will use `release` mode for builds. If a package
should be built in `debug` mode, it can be configured like so:
```

```nix
rustPlatform.buildRustPackage {
  /* ... */
  buildType = "debug";
}
```

> Obviously, the `checkPhase` will be ran in `debug`-mode as well in this case.

Reviewer: ```suggestion
In this scenario, the `checkPhase` will be ran in `debug` mode as well.
```

### Custom `build`/`install`-procedures

> Some packages may use custom scripts for building/installing, e.g. with a `Makefile`.
> In that case it's recommended to always override the `build-`/`install-`/`checkPhase`.
> 
> Otherwise, it may be possible that one of the internal steps fails because of the
> modified directory structure of `target/`.

Reviewer: ```suggestion
Some packages may use custom scripts for building/installing, e.g. with a `Makefile`.
In these cases, it's recommended to override the `buildPhase`/`installPhase`/`checkPhase`.

Otherwise, some steps may fail because of the modified directory structure of `target/`.
```
